File: FFRv2/app/src/main/java/com/devingotaswitch/youruserpools/RegisterUser.java
package com.devingotaswitch.youruserpools;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.appcompat.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser;
import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUserAttributes;
import com.amazonaws.mobileconnectors.cognitoidentityprovider.handlers.SignUpHandler;
import com.amazonaws.services.cognitoidentityprovider.model.CodeDeliveryDetailsType;
import com.amazonaws.services.cognitoidentityprovider.model.SignUpResult;
import com.devingotaswitch.ffrv2.R;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

public class RegisterUser extends AppCompatActivity {

    private EditText username;
    private EditText password;
    private EditText givenName;
    private EditText phone;
    private EditText email;

    private AlertDialog userDialog;
    private AlertDialog waitDialog;
    private String usernameInput;
    private String userPasswd;

    private static final String TAG = "RegisterUser";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register_user);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            // get back to main screen
            String value = extras.getString("TODO");
            if (value.equals("exit")) {
                onBackPressed();
            }
        }

        Toolbar toolbar = findViewById(R.id.toolbar_Register);
        toolbar.setTitle("");
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        TextView main_title = findViewById(R.id.signUp_toolbar_title);
        main_title.setText("Sign up");

        toolbar.setNavigationOnClickListener(v -> onBackPressed());
        init();
    }


    // This will create the list/form for registration
    private void init() {
        username = findViewById(R.id.editTextRegUserId);
        username.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegUserIdLabel);
                    label.setText(username.getHint());
                    username.setBackground(getDrawable(R.drawable.text_border_selector));
                }
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                TextView label = findViewById(R.id.textViewRegUserIdMessage);
                label.setText("");
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegUserIdLabel);
                    label.setText("");
                }
            }
        });
        //
        password = findViewById(R.id.editTextRegUserPassword);
        password.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if(s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegUserPasswordLabel);
                    label.setText(password.getHint());
                    password.setBackground(getDrawable(R.drawable.text_border_selector));
                }
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                TextView label = findViewById(R.id.textViewUserRegPasswordMessage);
                label.setText("");

            }

            @Override
            public void afterTextChanged(Editable s) {
                if(s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegUserPasswordLabel);
                    label.setText("");
                }
            }
        });
        //
        givenName = findViewById(R.id.editTextRegGivenName);
        givenName.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegGivenNameLabel);
                    label.setText(givenName.getHint());
                    givenName.setBackground(getDrawable(R.drawable.text_border_selector));
                }
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                TextView label = findViewById(R.id.textViewRegGivenNameMessage);
                label.setText("");
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegGivenNameLabel);
                    label.setText("");
                }
            }
        });
        //
        phone = findViewById(R.id.editTextRegPhone);
        phone.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegPhoneLabel);
                    label.setText(phone.getHint());
                    phone.setBackground(getDrawable(R.drawable.text_border_selector));
                }
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                TextView label = findViewById(R.id.textViewRegPhoneMessage);
                label.setText("");
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegPhoneLabel);
                    label.setText("");
                }
            }
        });
        //
        email = findViewById(R.id.editTextRegEmail);
        email.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegEmailLabel);
                    label.setText(email.getHint());
                    email.setBackground(getDrawable(R.drawable.text_border_selector));
                }
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                TextView label = findViewById(R.id.textViewRegEmailMessage);
                label.setText("");
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (s.length() == 0) {
                    TextView label = findViewById(R.id.textViewRegEmailLabel);
                    label.setText("");
                }
            }
        });

        Button signUp = findViewById(R.id.signUp);
        signUp.setOnClickListener(v -> {
            // Read user data and register
            CognitoUserAttributes userAttributes = new CognitoUserAttributes();

            usernameInput = username.getText().toString();
            if (usernameInput == null || usernameInput.isEmpty()) {
                TextView view = findViewById(R.id.textViewRegUserIdMessage);
                view.setText(username.getHint() + " cannot be empty");
                username.setBackground(getDrawable(R.drawable.text_border_error));
                return;
            }

            String userpasswordInput = password.getText().toString();
            userPasswd = userpasswordInput;
            if (userpasswordInput == null || userpasswordInput.isEmpty()) {
                TextView view = findViewById(R.id.textViewUserRegPasswordMessage);
                view.setText(password.getHint() + " cannot be empty");
                password.setBackground(getDrawable(R.drawable.text_border_error));
                return;
            }

            String userInput = givenName.getText().toString();
            if (userInput != null && userInput.length() > 0) {
                userAttributes.addAttribute(CUPHelper.getSignUpFieldsC2O().get("Name"), userInput);
            } else {
                TextView view = findViewById(R.id.textViewRegGivenNameMessage);
                view.setText(givenName.getHint() + " cannot be empty");
                givenName.setBackground(getDrawable(R.drawable.text_border_error));
                return;
            }

            String phoneNumber = phone.getText().toString();
            if (phoneNumber != null && phoneNumber.length() > 0) {
                userAttributes.addAttribute(CUPHelper.getSignUpFieldsC2O().get("Phone number"), phoneNumber);
            } else {
                TextView view = findViewById(R.id.textViewRegPhoneMessage);
                view.setText(phone.getHint() + " cannot be empty");
                phone.setBackground(getDrawable(R.drawable.text_border_error));
                return;
            }

            String emailAddress = email.getText().toString();
            if (emailAddress != null && emailAddress.length() > 0) {
                userAttributes.addAttribute(CUPHelper.getSignUpFieldsC2O().get("Email"), emailAddress);
            } else {
                TextView view = findViewById(R.id.textViewRegEmailMessage);
                view.setText(email.getHint() + " cannot be empty");
                email.setBackground(getDrawable(R.drawable.text_border_error));
                return;
            }

            showWaitDialog();

            CUPHelper.getPool().signUpInBackground(usernameInput, userpasswordInput,
                    userAttributes, null, signUpHandler);

        });
    }

    private final SignUpHandler signUpHandler = new SignUpHandler() {
        @Override
        public void onSuccess(CognitoUser user, SignUpResult signUpResult) {
            // Check signUpConfirmationState to see if the user is already confirmed
            closeWaitDialog();
            if (signUpResult.isUserConfirmed()) {
                // User is already confirmed
                showDialogMessage("Sign up successful!",usernameInput+" has been confirmed", true);
            }
            else {
                // User is not confirmed
                confirmSignUp(signUpResult.getCodeDeliveryDetails());
            }
        }

        @Override
        public void onFailure(Exception exception) {
            closeWaitDialog();
            TextView label = findViewById(R.id.textViewRegUserIdMessage);
            label.setText("Sign up failed");
            username.setBackground(getDrawable(R.drawable.text_border_error));
            showDialogMessage("Sign up failed", CUPHelper.formatException(exception),false);
        }
    };

    private void confirmSignUp(CodeDeliveryDetailsType cognitoUserCodeDeliveryDetails) {
        Intent intent = new Intent(this, SignUpConfirm.class);
        intent.putExtra("source","signup");
        intent.putExtra("name", usernameInput);
        intent.putExtra("destination", cognitoUserCodeDeliveryDetails.getDestination());
        intent.putExtra("deliveryMed", cognitoUserCodeDeliveryDetails.getDeliveryMedium());
        intent.putExtra("attribute", cognitoUserCodeDeliveryDetails.getAttributeName());

        startActivityForResult(intent, 10);
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 10) {
            if(resultCode == RESULT_OK){
                String name = null;
                if(data.hasExtra("name")) {
                    name = data.getStringExtra("name");
                }
                exit(name, userPasswd);
            }
        }
    }

    private void showDialogMessage(String title, String body, final boolean exit) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(title).setMessage(body).setNeutralButton("OK", (dialog, which) -> {
            try {
                userDialog.dismiss();
                if(exit) {
                    exit(usernameInput);
                }
            } catch (Exception e) {
                if(exit) {
                    exit(usernameInput);
                }
            }
        });
        userDialog = builder.create();
        userDialog.show();
    }

    private void showWaitDialog() {
        closeWaitDialog();
        waitDialog = new MaterialAlertDialogBuilder(this)
            .setTitle("Signing up...")
            .create();
        waitDialog.show();
    }

    private void closeWaitDialog() {
        try {
            waitDialog.dismiss();
        }
        catch (Exception e) {
            //
        }
    }

    private void exit(String uname) {
        exit(uname, null);
    }

    private void exit(String uname, String password) {
        Intent intent = new Intent();
        if (uname == null) {
            uname = "";
        }
        if (password == null) {
            password = "";
        }
        intent.putExtra("name", uname);
        intent.putExtra("password", password);
        setResult(RESULT_OK, intent);
        finish();
    }
}

File: FFRv2/app/src/main/java/com/devingotaswitch/youruserpools/RegisterUser.kt
package com.devingotaswitch.youruserpools

import android.content.DialogInterface
import android.content.Intent
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUser
import com.amazonaws.mobileconnectors.cognitoidentityprovider.CognitoUserAttributes
import com.amazonaws.mobileconnectors.cognitoidentityprovider.handlers.SignUpHandler
import com.amazonaws.services.cognitoidentityprovider.model.CodeDeliveryDetailsType
import com.amazonaws.services.cognitoidentityprovider.model.SignUpResult
import com.devingotaswitch.ffrv2.R
import com.devingotaswitch.youruserpools.CUPHelper.formatException
import com.devingotaswitch.youruserpools.CUPHelper.getSignUpFieldsC2O
import com.devingotaswitch.youruserpools.CUPHelper.pool
import com.devingotaswitch.youruserpools.SignUpConfirm
import com.google.android.material.dialog.MaterialAlertDialogBuilder

class RegisterUser : AppCompatActivity() {
    private var username: EditText? = null
    private var password: EditText? = null
    private var givenName: EditText? = null
    private var phone: EditText? = null
    private var email: EditText? = null
    private var userDialog: AlertDialog? = null
    private var waitDialog: AlertDialog? = null
    private var usernameInput: String? = null
    private var userPasswd: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register_user)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        val extras = intent.extras
        if (extras != null) {
            // get back to main screen
            val value = extras.getString("TODO")
            if (value == "exit") {
                onBackPressed()
            }
        }
        val toolbar = findViewById<Toolbar>(R.id.toolbar_Register)
        toolbar.title = ""
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        val main_title = findViewById<TextView>(R.id.signUp_toolbar_title)
        main_title.text = "Sign up"
        toolbar.setNavigationOnClickListener { v: View? -> onBackPressed() }
        init()
    }

    // This will create the list/form for registration
    private fun init() {
        username = findViewById(R.id.editTextRegUserId)
        username!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegUserIdLabel)
                    label.text = username!!.hint
                    username!!.background = ContextCompat.getDrawable(applicationContext,
                            R.drawable.text_border_selector)
                }
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val label = findViewById<TextView>(R.id.textViewRegUserIdMessage)
                label.text = ""
            }

            override fun afterTextChanged(s: Editable) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegUserIdLabel)
                    label.text = ""
                }
            }
        })
        //
        password = findViewById(R.id.editTextRegUserPassword)
        password!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegUserPasswordLabel)
                    label.text = password!!.hint
                    password!!.background = ContextCompat.getDrawable(applicationContext,
                            R.drawable.text_border_selector)
                }
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val label = findViewById<TextView>(R.id.textViewUserRegPasswordMessage)
                label.text = ""
            }

            override fun afterTextChanged(s: Editable) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegUserPasswordLabel)
                    label.text = ""
                }
            }
        })
        //
        givenName = findViewById(R.id.editTextRegGivenName)
        givenName!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegGivenNameLabel)
                    label.text = givenName!!.hint
                    givenName!!.background = ContextCompat.getDrawable(applicationContext,
                            R.drawable.text_border_selector)
                }
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val label = findViewById<TextView>(R.id.textViewRegGivenNameMessage)
                label.text = ""
            }

            override fun afterTextChanged(s: Editable) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegGivenNameLabel)
                    label.text = ""
                }
            }
        })
        //
        phone = findViewById(R.id.editTextRegPhone)
        phone!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegPhoneLabel)
                    label.text = phone!!.hint
                    phone!!.background = ContextCompat.getDrawable(applicationContext,
                            R.drawable.text_border_selector)
                }
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val label = findViewById<TextView>(R.id.textViewRegPhoneMessage)
                label.text = ""
            }

            override fun afterTextChanged(s: Editable) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegPhoneLabel)
                    label.text = ""
                }
            }
        })
        //
        email = findViewById(R.id.editTextRegEmail)
        email!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegEmailLabel)
                    label.text = email!!.hint
                    email!!.background = ContextCompat.getDrawable(applicationContext,
                            R.drawable.text_border_selector)
                }
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val label = findViewById<TextView>(R.id.textViewRegEmailMessage)
                label.text = ""
            }

            override fun afterTextChanged(s: Editable) {
                if (s.isEmpty()) {
                    val label = findViewById<TextView>(R.id.textViewRegEmailLabel)
                    label.text = ""
                }
            }
        })
        val signUp = findViewById<Button>(R.id.signUp)
        signUp.setOnClickListener { v: View? ->
            // Read user data and register
            val userAttributes = CognitoUserAttributes()
            usernameInput = username!!.text.toString()
            if (usernameInput == null || usernameInput!!.isEmpty()) {
                val view = findViewById<TextView>(R.id.textViewRegUserIdMessage)
                view.text = username!!.hint.toString() + " cannot be empty"
                username!!.background = ContextCompat.getDrawable(applicationContext,
                        R.drawable.text_border_error)
                return@setOnClickListener
            }
            val userpasswordInput = password!!.text.toString()
            userPasswd = userpasswordInput
            if (userpasswordInput.isEmpty()) {
                val view = findViewById<TextView>(R.id.textViewUserRegPasswordMessage)
                view.text = password!!.hint.toString() + " cannot be empty"
                password!!.background = ContextCompat.getDrawable(applicationContext,
                        R.drawable.text_border_error)
                return@setOnClickListener
            }
            val userInput = givenName!!.text.toString()
            if (userInput.isNotEmpty()) {
                userAttributes.addAttribute(getSignUpFieldsC2O()!!["Name"], userInput)
            } else {
                val view = findViewById<TextView>(R.id.textViewRegGivenNameMessage)
                view.text = givenName!!.hint.toString() + " cannot be empty"
                givenName!!.background = ContextCompat.getDrawable(applicationContext,
                        R.drawable.text_border_error)
                return@setOnClickListener
            }
            val phoneNumber = phone!!.text.toString()
            if (phoneNumber.isNotEmpty()) {
                userAttributes.addAttribute(getSignUpFieldsC2O()!!["Phone number"], phoneNumber)
            } else {
                val view = findViewById<TextView>(R.id.textViewRegPhoneMessage)
                view.text = phone!!.hint.toString() + " cannot be empty"
                phone!!.background = ContextCompat.getDrawable(applicationContext,
                        R.drawable.text_border_error)
                return@setOnClickListener
            }
            val emailAddress = email!!.text.toString()
            if (emailAddress.isNotEmpty()) {
                userAttributes.addAttribute(getSignUpFieldsC2O()!!["Email"], emailAddress)
            } else {
                val view = findViewById<TextView>(R.id.textViewRegEmailMessage)
                view.text = email!!.hint.toString() + " cannot be empty"
                email!!.background = ContextCompat.getDrawable(applicationContext,
                        R.drawable.text_border_error)
                return@setOnClickListener
            }
            showWaitDialog()
            pool!!.signUpInBackground(usernameInput, userpasswordInput,
                    userAttributes, null, signUpHandler)
        }
    }

    private val signUpHandler: SignUpHandler = object : SignUpHandler {
        override fun onSuccess(user: CognitoUser, signUpResult: SignUpResult) {
            // Check signUpConfirmationState to see if the user is already confirmed
            closeWaitDialog()
            if (signUpResult.isUserConfirmed) {
                // User is already confirmed
                showDialogMessage("Sign up successful!", "$usernameInput has been confirmed", true)
            } else {
                // User is not confirmed
                confirmSignUp(signUpResult.codeDeliveryDetails)
            }
        }

        override fun onFailure(exception: Exception) {
            closeWaitDialog()
            val label = findViewById<TextView>(R.id.textViewRegUserIdMessage)
            label.text = "Sign up failed"
            username!!.background = ContextCompat.getDrawable(applicationContext,
                    R.drawable.text_border_error)
            showDialogMessage("Sign up failed", formatException(exception), false)
        }
    }

    private fun confirmSignUp(cognitoUserCodeDeliveryDetails: CodeDeliveryDetailsType) {
        val intent = Intent(this, SignUpConfirm::class.java)
        intent.putExtra("source", "signup")
        intent.putExtra("name", usernameInput)
        intent.putExtra("destination", cognitoUserCodeDeliveryDetails.destination)
        intent.putExtra("deliveryMed", cognitoUserCodeDeliveryDetails.deliveryMedium)
        intent.putExtra("attribute", cognitoUserCodeDeliveryDetails.attributeName)
        startActivityForResult(intent, 10)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 10) {
            if (resultCode == RESULT_OK) {
                var name: String? = null
                if (data!!.hasExtra("name")) {
                    name = data.getStringExtra("name")
                }
                exit(name, userPasswd)
            }
        }
    }

    private fun showDialogMessage(title: String, body: String, exit: Boolean) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(title).setMessage(body).setNeutralButton("OK") { _: DialogInterface?, _: Int ->
            try {
                userDialog!!.dismiss()
                if (exit) {
                    exit(usernameInput)
                }
            } catch (e: Exception) {
                if (exit) {
                    exit(usernameInput)
                }
            }
        }
        userDialog = builder.create()
        userDialog!!.show()
    }

    private fun showWaitDialog() {
        closeWaitDialog()
        waitDialog = MaterialAlertDialogBuilder(this)
                .setTitle("Signing up...")
                .create()
        waitDialog!!.show()
    }

    private fun closeWaitDialog() {
        try {
            waitDialog!!.dismiss()
        } catch (e: Exception) {
            //
        }
    }

    private fun exit(uname: String?, password: String? = null) {
        var uname = uname
        var password = password
        val intent = Intent()
        if (uname == null) {
            uname = ""
        }
        if (password == null) {
            password = ""
        }
        intent.putExtra("name", uname)
        intent.putExtra("password", password)
        setResult(RESULT_OK, intent)
        finish()
    }

    companion object {
        private const val TAG = "RegisterUser"
    }
}